### PR TITLE
[FLINK-32029][core] Adds fallback error handling to FutureUtils.handleUncaughtException

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

I ran into an issue in FLINK-31773 where a `NullPointerException` was thrown by the error handling code (because the contender wasn't set, anymore). That one went unnoticed the the corresponding unit tests because we don't handle errors in the error handling code. This PR adds this code path.

## Brief change log

* Migrates `FutureUtilsTest` from JUnit4/Hamcrest to JUnit5/AssertJ (separate commit)
* Adds fallback error handling to `FutureUtils.handleUncaughtException`

## Verifying this change

* a helper method was added that allows configuring the fallback exception handler
* a test was added covering the scenario

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorExtension.java [hunk 1]
```diff
@@ -49,6 +58,20 @@ public T getExecutor() {
     public void afterAll(ExtensionContext context) throws Exception {
         if (executorService != null) {
             executorService.shutdown();
+
+            try {
```
Comment (pgaref) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorExtension.java:62]: Is this change related?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorResource.java [hunk 2]
```diff
@@ -47,6 +55,20 @@ public T getExecutor() {
     protected void after() {
         if (executorService != null) {
             executorService.shutdown();
+
```
Comment (pgaref) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorResource.java:58]: same as above

**Review by pgaref — APPROVED** [review]
Fallback FatalExitExceptionHandler LGTM.
Minor comment about some not super related changes

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorExtension.java (hunk 1 above)
Comment (XComp) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorExtension.java]: No, this change is not related. I was too lazy to create a Jira issue for that one and just added it as a minor hotfix commit. But you're right. It's too different to let it slip through. :-) I created FLINK-32039 to cover it.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorResource.java (hunk 2 above)
Comment (XComp) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorResource.java]: See [my comment above](https://github.com/apache/flink/pull/22544#discussion_r1188199532)

**Comment by XComp** [comment]
I dropped the Executor shutdown commit (this one is moved to FLINK-32039) and rebased the branch. Thanks for your review, @pgaref

**Review by snuyanzin — APPROVED** [review]
